presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/SymbolMapper.java

```diff
@@ -0,0 +1,112 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.planner.optimizations;
+
+import com.facebook.presto.metadata.Signature;
+import com.facebook.presto.sql.planner.Symbol;
+import com.facebook.presto.sql.planner.plan.AggregationNode;
+import com.facebook.presto.sql.planner.plan.PlanNode;
+import com.facebook.presto.sql.planner.plan.PlanNodeId;
+import com.facebook.presto.sql.tree.Expression;
+import com.facebook.presto.sql.tree.ExpressionRewriter;
+import com.facebook.presto.sql.tree.ExpressionTreeRewriter;
+import com.facebook.presto.sql.tree.FunctionCall;
+import com.facebook.presto.sql.tree.SymbolReference;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import static com.facebook.presto.util.ImmutableCollectors.toImmutableList;
+import static java.util.Objects.requireNonNull;
+
+public class SymbolMapper
+{
+    private final Map<Symbol, Symbol> mapping;
+
+    public SymbolMapper(Map<Symbol, Symbol> mapping)
+    {
+        this.mapping = ImmutableMap.copyOf(requireNonNull(mapping, "mapping is null"));
+    }
+
+    public Symbol map(Symbol symbol)
+    {
+        Symbol canonical = symbol;
+        while (mapping.containsKey(canonical) && !mapping.get(canonical).equals(canonical)) {
+            canonical = mapping.get(canonical);
+        }
+        return canonical;
+    }
+
+    public Expression map(Expression value)
+    {
+        return ExpressionTreeRewriter.rewriteWith(new ExpressionRewriter<Void>()
+        {
+            @Override
+            public Expression rewriteSymbolReference(SymbolReference node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+            {
+                Symbol canonical = map(Symbol.from(node));
+                return canonical.toSymbolReference();
+            }
+        }, value);
+    }
+
+    public AggregationNode map(AggregationNode node, PlanNode source, PlanNodeId nodeId)
```

Comment:
why do you need to pass `nodeId`, it would be better when `SymbolMapper` had `idAllocator`. Alterantively why not to reuse id from `node`